### ISSUE TITLE
Support even older Windows versions

### ISSUE THREAD

**Issue by RamonUnch** [RamonUnch/GreenPad#58]
For now GreenPad works fine on:
Windows NT3.1 build 404 (oldest build I tested)
Win3.1 + Win32s 1.15a

I tested Windows 3.1 + Win32s 1.1 (build 88) and the System freezes without other errors when I try to start GreenPad,
If I delete the .ini file it gets created again, so the lockup does not happen immediately.

**Comment by RamonUnch** [comment]
Actually I just needed to define NO_ASMTHUNK and the exe works on Win3.1+Win32s 1.1
I just need to try with even older Win32s versions.

**Comment by RamonUnch** [comment]
Seems useless to support Win32s beta because it does not support codepages and WideCharToMultiByte/MultiByteToWideCharare stubs.
https://github.com/RamonUnch/GreenPad/commit/0047c940646b355ec7dcccf753146343931c84e7

**Comment by roytam1** [comment]
my build crashes with GPF in chicago build 58s:
![image](https://user-images.githubusercontent.com/1192577/211747624-001e4ed9-8e8e-41de-9001-37fbdd735df1.png)

I wonder if I completely disabling status bar will help or not.

**Comment by roytam1** [comment]
yeah kind of working now, but it seems that WC2MB and/or MB2WC is failing:
![image](https://user-images.githubusercontent.com/1192577/211755154-df74c254-eacb-455a-9f3a-52ad8c06b88b.png)

I can type something but it returns empty lines.

**Comment by RamonUnch** [comment]
Probably GetCharWidthW and TextOutW are also not supported on old Chicago builds.

**Comment by RamonUnch** [comment]
For me it was enough to set `StatusBar=0` in the ini file

**Comment by roytam1** [comment]
> Probably GetCharWidthW and TextOutW are also not supported on old Chicago builds.

since I'm testing win32s builds not NT31 ANSI builds so this is not a problem here.

**Comment by roytam1** [comment]
for NT 340, I can make it opening Open File dialog but it looks quirk: 
![image](https://user-images.githubusercontent.com/1192577/212450141-36c6df8f-9802-4337-bce9-25d5499de918.png)

and Save File dialog can't be shown right now.

**Comment by roytam1** [comment]
changed dialog template font, open dialog works now:
![image](https://user-images.githubusercontent.com/1192577/212457654-50d795a0-ed7e-41e2-8ed8-07ef8352de39.png)

but Save File (As) still returns LastError 6 unless I remove OFN_ENABLEHOOK flag (but this is deal breaker)
![image](https://user-images.githubusercontent.com/1192577/212457661-f629769e-3372-4cd0-8ce5-ae0e2e7c25b7.png)

**Comment by RamonUnch** [comment]
Quite some progress indeed, congratulations!

**Comment by roytam1** [comment]
> but Save File (As) still returns LastError 6 unless I remove OFN_ENABLEHOOK flag (but this is deal breaker)
> ![image](https://user-images.githubusercontent.com/1192577/212457661-f629769e-3372-4cd0-8ce5-ae0e2e7c25b7.png)

for further testing, I used its SDK samples\multipad adding OFN_ENABLEHOOK for both GetOpenFilename and GetSaveFilename with a stub ofnhook function which just returns FALSE, such save hook works on later versions but not in 340.
I wonder that 340's comdlg32.dll may need some binary patching.

**Comment by RamonUnch** [comment]
How do you get status bar on NT build 340?

**Comment by RamonUnch** [comment]
Actually copying comctl32.dll from build 404 did the trick for me!

**Comment by RamonUnch** [comment]
For me on NT3.10.340, I can barely use the open dialog. After a fresh restart it displays fine, but then
1) If I do Ctrl+O then I press esc,/cancel I get insta crash.
2) When I select a file everything is OK, however later instances of GreenPad get a glitched Open dialog that aslo crashed on cancel.
So I need a full restart. login/logout does not do it.
Notepad does not have the problem though.

**Comment by roytam1** [comment]
> 2\. Notepad does not have the problem though.

because notepad doesn't use common dialog templates and hooks.

**Comment by RamonUnch** [comment]
So I guess we have to drop the hook and resolve to the simple open dialog.
The codepage can be selected later via the reopen dialog.
For saving that is a bigger issue. So we should allow to change current cs even on unsaved files, so that it gets used for saving.

**Comment by roytam1** [comment]
> So I guess we have to drop the hook and resolve to the simple open dialog.

it has no problem on open dialog (I can't test via hotkeys, but via menu is no problem here)

**Comment by RamonUnch** [comment]
> it has no problem on open dialog (I can't test via hotkeys, but via menu is no problem here)

You are right, I just had problems because I was using "MS Shell Dlg" font, with "Helv" I get no problems.

**Comment by roytam1** [comment]
I got it finally!
![image](https://user-images.githubusercontent.com/1192577/213081374-720160b0-2ec3-4ac6-9542-ef8da4d676a4.png)

NT 340's comdlg32.dll wants ofnHook returning TRUE when processing WM_INITDIALOG.

**Comment by RamonUnch** [comment]
Amazing, I will give it a try.

**Comment by RamonUnch** [comment]
According to the doc, we should return TRUE for the `WM_INITDIALOG` processing even (i guess) for the open dialog even though it works with return FALSE.

DOC: for OFNHookProcOldStyle
>**Return Values**
>If the hook procedure returns zero, the default dialog box procedure processes the message.
If the hook procedure returns a nonzero value, the default dialog box procedure ignores the message.
...

 > The default dialog box procedure processes the WM_INITDIALOG message before passing it to the hook procedure. For all other messages, the hook procedure receives the message first. Then, the return value of the hook procedure determines whether the default dialog procedure processes the message or ignores it.

For the new style dialog it is also safer to return nonzero value from WM_INITDIALOG:
>If you provide a hook procedure for an Explorer-style common dialog box, the system creates a dialog box that is a child of the default dialog box. The hook procedure acts as the dialog procedure for the child dialog. This child dialog is based on the template you specified in the OPENFILENAME structure, or it is a default child dialog if no template is specified. The child dialog is created when the default dialog procedure is processing its WM_INITDIALOG message. After the child dialog processes its own WM_INITDIALOG message, the default dialog procedure moves the standard controls, if necessary, to make room for any additional controls of the child dialog. The system then sends the CDN_INITDONE notification message to the hook procedure.

So I guess this return TRUE should always apply.

**Comment by roytam1** [comment]
remaining systems:
* NT Build < 329/340
  - since EXE format differences from later versions, there is no good way to support it
* Chicago Build 58s and Win32s < 1.10.88
  - WC2MB and MB2WC are stubs
* Win32s < 1.10.88 (cont.)
  - CreateFileMapping() fails without setting LastError

**Comment by RamonUnch** [comment]
I did not test GreenPad under OS/2 warp.

It should be possible via two ways:
1) [Installing Win32s under WIN-OS2](https://www.os2world.com/wiki/index.php/Installing_Win32s_in_Win-OS/2)
2) Using [Odin](http://odin.netlabs.org/en/site/index.xml)

The odds are high everything works fine already but I will ll need to test this when I get a working OS/2 environment.
OS/2 came back to my memory when I heard about [ArcaOS](https://www.arcanoae.com/arcaos/). This latter should also have a native support for Win32 programs, it seems to be a continuation of [eComStation](https://www.ecomstation.com/) that stopped development in 2011.

**Comment by RamonUnch** [comment]
From quick testing it seems that Win32s build of GreenPad works fine on WIN-OS2+Win32s1.25a (with comctl32.dll from beta) under OS/2 Warp 4.51.

**Comment by roytam1** [comment]
> 2\. Using [Odin](http://odin.netlabs.org/en/site/index.xml)

my old odin install is still working.
![image](https://user-images.githubusercontent.com/1192577/214014045-b12897bf-2b3b-4f04-83a2-74a877975bfd.png)
odin can only work with ANSI versions, and it work best with my NT31-ansi version.

**Comment by RamonUnch** [comment]
Wow, that's so beautiful!
Which Windows version does Odin identifies itself as?

**Comment by roytam1** [comment]
> Which Windows version does Odin identifies itself as?

XP.
![image](https://user-images.githubusercontent.com/1192577/214018027-a276bb33-ccaf-4563-890e-95aa9ec511b5.png)

**Comment by RamonUnch** [comment]
Interesting to report that since it does not support proper Unicode...

**Comment by RamonUnch** [comment]
I took a glance at ReactOS source and It seems the DC is released on ReleaseDC()/EndPaint() even when the CS_OWNDC style is specified,
the CS_OWNDC style is checked however to create a DC when calling CreateWindow which is good, I am unsure however how it is handled elsewere, But I need to check if the OWNDC version of GreenPad would work if I remove the EndPaint() and use ValidateRect() instead.

**Comment by roytam1** [comment]
> I took a glance at ReactOS source and It seems the DC is released on ReleaseDC()/EndPaint() even when the CS_OWNDC style is specified,

but ros winetest doesn't show that:
https://github.com/reactos/reactos/blob/master/modules/rostests/winetests/user32/dce.c
https://jira.reactos.org/browse/CORE-18800?focusedCommentId=138102&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-138102

**Comment by RamonUnch** [comment]
Well at least if I replace BeginPaint/EndPaint with GetUpdateRegion/ValidateRect in the old greenpad, I no longer have huge problems, and the Window content can still get repainted even after minimizing/restoring a lot the window. There are still drawing problems though.
Hwever I do not know if that proves anything because in the end if we remove Begin/end paint we only call `GetDC()` once and then never release it, so it cannot change, However I checked that every call to GetDC() gives the same than the one that was saved in the Painter, and it is the case on both ReactOS and Server2003.
So if we never call ReleaseDC or EndPaint() it seems to go mostly fine.

**Comment by roytam1** [comment]
opening settings dialog cause a app freeze in NT 340 in my tree, does it happen in yours?
![image](https://user-images.githubusercontent.com/1192577/214750335-cfd6fc90-ad9a-4bd5-bc1d-30803fa6e5f7.png)

**Comment by RamonUnch** [comment]
I have no problems on NT3.10.340 with settings dialog, both using the Unicode beta build or win32s beta:
![image](https://user-images.githubusercontent.com/74856804/214793193-51078645-ca94-4abd-bada-29cee4488628.png)

**Comment by roytam1** [comment]
ah, I just forget to pass TARGET_VER to RC, fixed my side. :)

**Comment by RamonUnch** [comment]
Windows NT 3.10.340 cannot set the text on the help dialog, is it just me?
The rest works fine...
![image](https://user-images.githubusercontent.com/74856804/225573821-1b9b4bd4-2392-4568-865e-a7029d807c69.png)

**Comment by RamonUnch** [comment]
Forget it, it is because of big font, If I set font size back to 8 then everything is fine.
In all cases the SendDlgItemMessage() returns true and sets lasterror to 1447 (ERROR_NO_SCROLLBARS)

**Comment by roytam1** [comment]
ported changes into my nt31 tree now:
![image](https://user-images.githubusercontent.com/1192577/227697461-fae0aaed-93f3-48ea-b4f3-f885c2e3fe63.png)

**Comment by roytam1** [comment]
more screenshots:
![image](https://user-images.githubusercontent.com/1192577/227700129-5c32d3ce-39e6-4285-89f9-1eae9e594b52.png)
![image](https://user-images.githubusercontent.com/1192577/227700049-b3ef326b-9d66-4c45-a659-387c9614f89c.png)

![image](https://user-images.githubusercontent.com/1192577/227700208-7f3af2a8-4b54-4d5c-b342-299524ebe82d.png)
![image](https://user-images.githubusercontent.com/1192577/227700233-c4394dd0-d30b-478e-9eda-56069de74a89.png)

![image](https://user-images.githubusercontent.com/1192577/227700459-e95d7c24-9047-4806-8577-a0ca417ca481.png)

**Comment by RamonUnch** [comment]
That is beautiful indeed!
I also tried to run greenpad on Win32s beta 32, and it runs and kinda work. The problem is that the GetCharWidth does not work (I think). I did not invest an time into it for now.

**Comment by roytam1** [comment]
BTW I tried to dedup and correct some open/save dialog codes and logics:
https://github.com/roytam1/rtoss/compare/edfb61d49cd21631d87396f95a9f1cbbb14528ed...b1ebfb30ee5a0b0ba49c6ca528fcb47e8b0a04d9

**Comment by RamonUnch** [comment]
I will import your changes.
A change I made is to remove extra checks for the First if after dialog opening failed.
As soon as you have the dlgEvenOpened_ variable there is no need for any extra checks other than `OFN_EXPLORER` before trying again with old dialog.

**Comment by RamonUnch** [comment]
Thanks for your browsers builds btw,
I noted an issue, even with the latest IceApe-UXP build:
Sometime, after posting in a GitHub thread and reloading the page (not 100% sure on how to reproduce), The whole thread gets dplicated: ie:

Initial post:
Post1
Post2
Post3
Post1
Post2
Post3

I do have the problem right now, You can see that tin the middle of the page the whole thread repeats with my first post abiut disabling asmthubk :
![image](https://user-images.githubusercontent.com/74856804/227708478-9f353d06-4855-4853-91f7-1e2182cdcf83.png)

Not sure if you ever experienced this issue.
I did experience it also using palefill before WebComponents were enabled in UXP

**Comment by roytam1** [comment]
> Not sure if you ever experienced this issue.

didn't notice this, maybe palefill and/or MC's WebComponents/ShadowRoot implementations are not good enough for github.

**Comment by RamonUnch** [comment]
I will try to find a reproducible way to trigger it with a clean profile on NM28, and I will report on the MSFN thread. Excuse me for this off-topic tangent.

**Comment by roytam1** [comment]
![image](https://github.com/RamonUnch/GreenPad/assets/1192577/ddbf4b6d-98c7-46e8-9d6d-18a724cefa93)
so my tree builds on AXP32, buildable for AXP64 but not tested.

**Comment by RamonUnch** [comment]
Wow that is cool, MLang is not really needed when you got chardet.dll.
I guess you are not using the internal memory manager either because it ignores all alignment rules.
Should not be hard to fix though, rounding up 4bytes the allocated size should be enough I conjecture.

**Comment by roytam1** [comment]
I added a way to detect Win9x KernelEx in my tree:
![image](https://github.com/RamonUnch/GreenPad/assets/1192577/54ad0677-87b4-4f93-9dd8-661d45050b15)

related commit: https://github.com/roytam1/rtoss/commit/0066039a247a6b310f06e2d1381d462e53a9f988

**Comment by RamonUnch** [comment]
nice trick, indeed on Win9x the system directory is does not end with 32,
Would there not be a false positive on 64 bit windows were the System32 is for 64 bits dlls and SysWow64 is for 32 bits dlls?

**Comment by roytam1** [comment]
> Would there not be a false positive on 64 bit windows were the System32 is for 64 bits dlls and SysWow64 is for 32 bits dlls?

thats why you need to call `Wow64DisableWow64FsRedirection` to disable WoW filesystem redirection. otherwise it always return system32 for compatibility.

**Comment by roytam1** [comment]
BTW tried on NT 3.1 Build 311, and it doesn't start.
I think its PE format may be still immature as QuickView can't see its DLL dependency. On Build 347 it works like on Build 340.

**Comment by roytam1** [comment]
just upgraded one of my win10ltsc vm to win11ltsc, and `RtlGetNtVersionNumbers` still works.
![image](https://github.com/user-attachments/assets/830f933d-1d28-402a-ad4f-bebe839e54b7)

**Comment by roytam1** [comment]
haven't retest on win32s for a while, retested today and found that sometimes open dialog doesn't init properly in win32s in first time:
<img width="808" height="627" alt="Image" src="https://github.com/user-attachments/assets/f5e5ffed-4706-4a16-9fc4-999482bf4901" />

and of course the old bug that file list is empty unless changing to all files and changing back.
<img width="808" height="627" alt="Image" src="https://github.com/user-attachments/assets/1d41957e-d01a-4b53-959c-06a0da84ead1" />
<img width="808" height="627" alt="Image" src="https://github.com/user-attachments/assets/cde01f3e-a6d7-47e2-946b-52c3b36b0d95" />

and finally I can test AXP64 with a forked version of qemu that emulates dec alphaserver es40:
<img width="808" height="651" alt="Image" src="https://github.com/user-attachments/assets/8d4af913-477b-48c3-bcf7-b4097f3381de" />
